api/api-shared-v2/src/main/java/com/thoughtworks/go/apiv2/shared/representers/EnvironmentVariableRepresenter.java

```diff
@@ -0,0 +1,86 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.thoughtworks.go.apiv2.shared.representers;
+
+import com.thoughtworks.go.api.base.OutputWriter;
+import com.thoughtworks.go.api.representers.ErrorGetter;
+import com.thoughtworks.go.api.representers.JsonReader;
+import com.thoughtworks.go.api.util.HaltApiResponses;
+import com.thoughtworks.go.apiv2.shared.exception.InvalidGoCipherTextRuntimeException;
+import com.thoughtworks.go.config.EnvironmentVariableConfig;
+import com.thoughtworks.go.config.EnvironmentVariablesConfig;
+import com.thoughtworks.go.security.CryptoException;
+
+import java.util.HashMap;
+import java.util.Optional;
+
+
+public class EnvironmentVariableRepresenter {
+    public static void toJSON(OutputWriter outputWriter, EnvironmentVariableConfig environmentVariableConfig) {
+        if (!environmentVariableConfig.errors().isEmpty()) {
+            outputWriter.addChild("errors", errorWriter -> {
+                HashMap<String, String> mapping = new HashMap<>();
+                mapping.put("encryptedValue", "encrypted_value");
+                new ErrorGetter(mapping).toJSON(outputWriter, environmentVariableConfig);
+            });
+        }
+
+        outputWriter
+                .add("secure", environmentVariableConfig.isSecure())
+                .add("name", environmentVariableConfig.getName());
+        addValue(outputWriter, environmentVariableConfig);
+    }
+
+    public static EnvironmentVariablesConfig fromJSONArray(JsonReader jsonReader) {
+        EnvironmentVariablesConfig variables = new EnvironmentVariablesConfig();
+        jsonReader.readArrayIfPresent("environment_variables", environmentVariables -> {
+            environmentVariables.forEach(variable -> {
+                variables.add(EnvironmentVariableRepresenter.fromJSON(new JsonReader(variable.getAsJsonObject())));
+            });
+        });
+        return variables;
+    }
+
+    public static EnvironmentVariableConfig fromJSON(JsonReader jsonReader) {
+        String name = jsonReader.getString("name");
+        Boolean secure = jsonReader.optBoolean("secure").orElse(false);
+        Optional<String> optValue = jsonReader.optString("value");
+        Optional<String> optEncValue = jsonReader.optString("encrypted_value");
+
+        if (!optValue.isPresent() && !optEncValue.isPresent()) {
+            HaltApiResponses.haltBecauseOfReason("Environment variable must contain either 'value' or 'encrypted_value'");
+        }
+
+        String value = secure ? optValue.orElse(null) : jsonReader.getString("value");
```

Comment:
Doesn't matter if it is secure or not at this point, deserialize method checks for it, all that is needed to be done is pass the string for "value" if present, else null. Why check for secure?

Reply:
You are right this check is there in deserialize, I will remove it from here.

Reply:
I think this check will have to be there, because if incoming request is something like this 
```
{
  "secure": false,
  "encrypted_value": "..."
}
```
then it should fail with 422 (`jsonReader.getString("value")`), because a `secure: false` request  should not have `encrypted_value`, it should have `value`.  This is similar to what we are doing in other places, when we are de-serializing environment variable.

Reply:
This is handled by https://github.com/gocd/gocd/blob/master/config/config-api/src/main/java/com/thoughtworks/go/config/EnvironmentVariableConfig.java#L298

Reply:
> This is handled by https://github.com/gocd/gocd/blob/master/config/config-api/src/main/java/com/thoughtworks/go/config/EnvironmentVariableConfig.java#L298 ?

This code is adds it as one of the errors but doesn't cause the request to fail with 422. 

I am keeping following line as the request should fail with 422.

`String value = secure ? optValue.orElse(null) : jsonReader.getString("value");`